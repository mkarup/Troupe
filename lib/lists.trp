let fun map f list =
    case list of
       [] => []
      |  (x::xs) => (f x) :: (map f xs)

    fun mapi f list =
    let fun mapj j ls = 
            case ls of [] => []
            |  (x::xs) => (f (j,x)) :: (mapj (j+1) xs)
    in mapj 0 list
    end
    
    

    fun foldl f y [] = y
      | foldl f y (x::xs) = foldl f (f (x,y)) xs

    fun length [] = 0
      | length (x::xs) = 1 + length xs

    fun reverse ls =
      let fun rev [] a = a
            | rev (x::xs) a = rev xs (x::a)

      in rev ls []
      end

    fun range n =
     let fun rng 0 = []
          |  rng k = k :: (rng (k-1))
      in reverse (rng n)
     end

    fun append l1 l2 =
        case l1 of
            [] => l2
          | x::l => x::(append l l2)

    fun nth (x::l) 1 = x
      | nth (x::l) n = nth l (n - 1)

    fun lookup list key default =
      case list of [] => default
                 | ((x, v):: xs) =>
                   if x = key then v
                   else lookup xs key default

    fun elem x [] = false 
      | elem x (y::ys) = if x = y then true else elem x ys

    fun partition f ls = 
      let fun partition_aux  (a,b) [] = (reverse a, reverse b ) 
            | partition_aux  (a,b) (x::xs) =
                        if f x then partition_aux (x::a, b) xs 
                               else partition_aux (a, x::b) xs
            
      
      in  partition_aux ([],[]) ls 
      end   

in
    [ ("map",     map)
    , ("mapi",    mapi)
    , ("foldl",   foldl)
    , ("range",   range)
    , ("reverse", reverse)
    , ("lookup",  lookup) 
    , ("elem",    elem)
    , ("length",  length)
    , ("append",  append)
    , ("partition", partition) 
    , ("nth", nth)
    ]
end
