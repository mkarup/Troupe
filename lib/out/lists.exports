map
mapi
foldl
range
reverse
lookup
elem
length
append
partition
nth