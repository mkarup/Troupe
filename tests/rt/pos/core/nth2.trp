import lists
let val ls = ["ls", 3, 1/2, {r = 10, p = 7}]
    val string = nth ls 1
    val int = nth ls 2
    val float = nth ls 3
    val record = nth ls 4 
in (getType string = "string") andalso (getType int = "number") andalso (getType float = "number") andalso (getType record = "record")
end
