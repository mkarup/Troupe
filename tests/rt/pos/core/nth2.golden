2024-06-09T16:28:52.670Z [RTM] [32minfo[39m: Skipping network creation. Observe that all external IO operations will yield a runtime error.
>>> Main thread finished with value: true@{}%{}
